lib/src/summation/counter.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// counter.dart
+// A flexible counter implementation.
+//
+// 2024 August 26
+// Author: Max Korbel <max.korbel@intel.com>
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+import 'package:rohd_hcl/src/summation/summation_base.dart';
+
+/// Keeps a count of the running sum of any number of sources with optional
+/// configuration for widths, saturation behavior, and restarting.
+class Counter extends SummationBase {
+  /// The output value of the counter.
+  Logic get count => output('count');
+
+  /// Creates a counter that increments according to the provided [interfaces].
+  ///
+  /// The [width] can be either explicitly provided or inferred from other
+  /// values such as a [maxValue], [minValue], or [resetValue] that contain
+  /// width information (e.g. a [LogicValue]), or by making it large enough to
+  /// fit [maxValue], or by inspecting widths of [interfaces]. There must be
+  /// enough information provided to determine the [width].
+  ///
+  /// If no [maxValue] is provided, one will be inferred by the maximum that can
+  /// fit inside of the [width].
+  ///
+  /// The [restart] input can be used to restart the counter to a new value, but
+  /// also continue to increment in that same cycle. This is distinct from
+  /// [reset] which will reset the counter, holding the [count] at [resetValue].
```

Comment:
Do you want to define the saturates feature and the conditions like overflow and equalsMax?

Reply:
A good idea, adding, thanks